### PR TITLE
Stm32 lptim fix

### PR DESCRIPTION
**Pull request by jdascenzio**
Solves several problems on stm32 lptimer

### CODE REVIEW

File: drivers/timer/stm32_lptim_timer.c [hunk 1]
```diff
@@ -47,32 +47,21 @@ static void lptim_irq_handler(struct device *unused)
...
 		/* do not change ARR yet, z_clock_announce will do */
 		LL_LPTIM_ClearFLAG_ARRM(LPTIM1);
 
-		/* increase the total nb of lptim count
+		/* increase the total nb of autoreload count
 		 * used in the z_timer_cycle_get_32() function.
 		 * Reading the CNT register gives a reliable value
 		 */
-		accumulated_lptim_cnt += lp_time;
+		u32_t auotreload = LL_LPTIM_GetAutoReload(LPTIM1) + 1;
```
Comment (cburlacu) [inline drivers/timer/stm32_lptim_timer.c:57]: NIT: typo

Comment (jdascenzio) [reply drivers/timer/stm32_lptim_timer.c]: Fix thanks!

**Comment by FRASTM** [comment]
thanks, 
Considering this PR with high interest for fixing the https://github.com/zephyrproject-rtos/zephyr/issues/20991

**Comment by erwango** [comment]
> thanks,
> Considering this PR with high interest for fixing the #20991

@FRASTM so do you approve or do you need more time to review ?

**Comment by FRASTM** [comment]
This is valid  in a "ticked" mode, when CONFIG_TICKLESS_KERNEL is "n"

**Comment by jdascenzio** [comment]
I found these bugs by observing a time difference between the RTC and the lptimer when it both uses the LSI clock.

**Comment by FRASTM** [comment]
Did you check the time difference in case of LSE, if setting ` CONFIG_SYS_CLOCK_TICKS_PER_SEC + 1;`

**Comment by jdascenzio** [comment]
Yes, I also observe a time difference with LSE without this correction.
I use 8192 for system tick.  We must have `(LPTIM_CLOCK % CONFIG_SYS_CLOCK_TICKS_PER_SEC) == 0' to not lose some clocks when calculating:
```
static void lptim_irq_handler(struct device *unused)
{
....
u32_t dticks = (autoreload
                             * CONFIG_SYS_CLOCK_TICKS_PER_SEC)
			     / LPTIM_CLOCK;
....
}
```

**Comment by FRASTM** [comment]
Testing on nucleo_wb55rg with CONFIG_STM32_LPTIM_CLOCK_LSE and CONFIG_SYS_CLOCK_TICKS_PER_SEC=8192 gives an error on test tests/kernel/timer/timer_api
A lower value CONFIG_SYS_CLOCK_TICKS_PER_SEC=4096  will pass this test.
However tests/kernel/sched/schedule_api/  still fails if CONFIG_TIMESLICING=y

**Comment by aurel32** [comment]
> Yes, I also observe a time difference with LSE without this correction.
> I use 8192 for system tick. We must have `(LPTIM_CLOCK % CONFIG_SYS_CLOCK_TICKS_PER_SEC) == 0' to not lose some clocks when calculating:
> 
> ```
> static void lptim_irq_handler(struct device *unused)
> {
> ....
> u32_t dticks = (autoreload
>                              * CONFIG_SYS_CLOCK_TICKS_PER_SEC)
> 			     / LPTIM_CLOCK;
> ....
> }
> ```

Yes, I think that's the reason why on the nRF52, we have:
```
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=32768
CONFIG_SYS_CLOCK_TICKS_PER_SEC=32768
```
Even if the ticks per sec value seems high, for a tickless kernel it is mostly defines the accuracy of the clock.

**Comment by jdascenzio** [comment]
@FRASTM Why backup domain is reset in z_clock_driver_init? It removes my RTC config...

**Comment by FRASTM** [comment]
The sequence of setting the LSE as clock source is coming from stm32 applications.
This is not the case with LSI.

**Comment by jdascenzio** [comment]
@RASWe don't need to reset

> The sequence of setting the LSE as clock source is coming from stm32 applications.
> This is not the case with LSI.

**Comment by jdascenzio** [comment]
> The sequence of setting the LSE as clock source is coming from stm32 applications.
> This is not the case with LSI.

We don't need to reset backup domain to enable and set LSE for lptime:
```
Peripherals independent clock configuration register (RCC_CCIPR)

Bits 19:18 LPTIM1SEL[1:0]: Low power timer 1 clock source selection
These bits are set and cleared by software to select the LPTIM1 clock source.
00: PCLK selected as LPTIM1 clock
01: LSI clock selected as LPTIM1 clock
10: HSI16 clock selected as LPTIM1 clock
11: LSE clock selected as LPTIM1 clock
```
But this is the case for rtc only if the source change:
```
Backup domain control register (RCC_BDCR)

Bits 9:8 RTCSEL[1:0]: RTC clock source selection
Set by software to select the clock source for the RTC. Once the RTC clock source has been
selected, it cannot be changed anymore unless the Backup domain is reset, or unless a
failure is detected on LSE (LSECSSD is set). The BDRST bit can be used to reset them.
00: No clock
01: LSE oscillator clock used as RTC clock
10: LSI oscillator clock used as RTC clock
11: HSE oscillator clock divided by 32 used as RTC clock
```
So, in my opinion, we must remove the call to ```LL_RCC_ForceBackupDomainReset()``` and ```LL_RCC_ReleaseBackupDomainReset()``` here to avoid conflict with RTC counter.

**Comment by FRASTM** [comment]
This sequence is given for LSE clock source 
	LL_PWR_EnableBkUpAccess();
	LL_RCC_ForceBackupDomainReset();
	LL_RCC_ReleaseBackupDomainReset();
and recommended by the STM32Cube for enabling the LSE.

**Comment by jdascenzio** [comment]
In HAL_RCCEx_PeriphCLKConfig The backup domain is only reset if RTC clock source changed

```
      /* Reset the Backup domain only if the RTC Clock source selection is modified from default */
      tmpregister = READ_BIT(RCC->BDCR, RCC_BDCR_RTCSEL);

      if((tmpregister != RCC_RTCCLKSOURCE_NONE) && (tmpregister != PeriphClkInit->RTCClockSelection))
      {
        /* Store the content of BDCR register before the reset of Backup Domain */
        tmpregister = READ_BIT(RCC->BDCR, ~(RCC_BDCR_RTCSEL));
        /* RTC Clock selection can be changed only if the Backup Domain is reset */
        __HAL_RCC_BACKUPRESET_FORCE();
        __HAL_RCC_BACKUPRESET_RELEASE();
        /* Restore the Content of BDCR register */
        RCC->BDCR = tmpregister;
      }

```

**Comment by FRASTM** [comment]
OK, this can be removed from the z_clock_driver_init function

**Comment by jdascenzio** [comment]
I create a commit which remove ```LL_RCC_ForceBackupDomainReset()``` and ```LL_RCC_ReleaseBackupDomainReset()```
We probably have to do something similar for RTC counter driver (counter_ll_stm32_rtc)

**Comment by erwango** [comment]
> I create a commit which remove `LL_RCC_ForceBackupDomainReset()` and `LL_RCC_ReleaseBackupDomainReset()`
> We probably have to do something similar for RTC counter driver (counter_ll_stm32_rtc)

@jdascenzio, To get this merged faster, I encourage you to create a dedicated PR, so you don't have to wait lptim fixes to be accepted. Also please create a matching issue as only bug fixes for known issues are being merged during the release process.

**Comment by carlescufi** [comment]
@jdascenzio will you please create a PR to close #20991 separate from this one? That way we can lower the medium priority bug count to release 2.3.0

**Comment by jdascenzio** [comment]
> > I create a commit which remove `LL_RCC_ForceBackupDomainReset()` and `LL_RCC_ReleaseBackupDomainReset()`
> > We probably have to do something similar for RTC counter driver (counter_ll_stm32_rtc)
> 
> @jdascenzio, To get this merged faster, I encourage you to create a dedicated PR, so you don't have to wait lptim fixes to be accepted. Also please create a matching issue as only bug fixes for known issues are being merged during the release process.

I create this PR #25576.

**Pull request closed without merging**